### DISCUSSION TITLE
wpcap dll not found

### DISCUSSION THREAD

**Discussion by GordonSimon** [chmorgan/sharppcap#489] [Q&A] [answered]
Exception thrown: 'System.DllNotFoundException' in SharpPcap.dll
An unhandled exception of type 'System.DllNotFoundException' occurred in SharpPcap.dll
Unable to load DLL 'wpcap' or one of its dependencies: The specified module could not be found. (0x8007007E)

Trying examples from PacketDotNet and SharpPcap.
Using Visual Studio 2022 on Windows. Loaded .NET 8 SDK, but creating .NET 7 console projects. Using NuGet Packet Manager to load { PacketDoNet V1.4.8-pre18, SharpPcap V6.2.5 }. Copy & paste examples into program.cs, for example => var devices = CaptureDeviceList.Instance; Can't get anything to execute with exception (although this line executes => var ver = SharpPcap.Pcap.SharpPcapVersion;)

Why is wpcap dll missing ? Am I supposed to locate the DLL and copy to my bin folder ? Am I using valid packages (i.e. V1.1.4.8-pre18)  ?

**Comment by GordonSimon** [comment]
Hmm - repeated the same exercise on a MAC; looks like it does not use the PacketDotNet V1.4.8-pre18 and so don't get that error. I will try to find PacketDotNet V1.4.7 to see if my problem gets fixed

**Comment by GordonSimon** [comment]
Changing to PacketDotNet V.1.4.7 on Windows PC, did not solve the problem.

I see there is a note about WinPcap deprecated - perhaps my problem is related to the deprecation; but I don't know a solution.

WinPcap has been deprecated
We recommend switching to LibPcapLiveDevice

**Comment by kayoub5** [comment, marked as answer]
@GordonSimon you need to install Npcap or WinPcap on your system to use sharppcap

**Reply by GordonSimon** [reply]
Thanks for your quick help.  I did install NPCap and I was able to get the sample to work.
Perhaps to complete this discussion, I can get confirmation on the following ...

[1] After loading NPCap, I still was not able to compile the sample due to the following ...

  var p = Packet.ParsePacket(rawCapture.GetLinkLayers(), rawCapture.Data);
  CS1061: 'RawCapture' does not contain a definition for GetLinkLayers ...
 
I replaced with rawCapture.LinkLayerType. Does this change look like the correct solution ?
  var p = Packet.ParsePacket(rawCapture.LinkLayerType, rawCapture.Data);


[2] Also - do you know why I need NPCap on Windows PC, but not on a MAC. I guess MAC has the wpcap.dll (or something similar) as part of the O/S ? (Maybe because MAC is "linux" like) ?

**Reply by kayoub5** [reply]
* Where did you find "Packet.ParsePacket(rawCapture.GetLinkLayers(), rawCapture.Data);" ?
* MAC have "libpcap"

**Reply by GordonSimon** [reply]
Hmm - I guess I took it from PacketDotnet. The Capture sample on SharpPCap referred to .NET 6, and what I thought was basically the same example on PacketDotnet referred to .NET 7.; I see now that only the PacketDotnet has the GetLinkLayer() reference. 

Indeed, the SharpPcap shows the code that I guess at "rawCapture.LinkLayerTYpe".

PacketDotNet example
https://github.com/dotpcap/packetnet/blob/master/Examples/CapturingAndParsingPackets/Main.cs

                var p = Packet.ParsePacket(rawCapture.GetLinkLayers(), rawCapture.Data);


SharpPcap
https://github.com/dotpcap/sharppcap/blob/master/Examples/CreatingCaptureFile/Program.cs

                var packet = PacketDotNet.Packet.ParsePacket(rawPacket.LinkLayerType, rawPacket.Data);

Again - thanks for all your help.

**Reply by GordonSimon** [reply]
Regarding libpcap ...
I have a "using" for SharpPcap.libpcap. I assumed that would provide libpcap from the SharpPcap assembly ?
Also - the greyed out text for the using suggests it is not referenced (so not sure I follow the reason or need for using SharpPcap.libpcap, even after loading NPcap). I am able to proceed with my objectives, so if you don't have time to look into this question - no problem. Again, I took this from examples on the GitHub.

![image](https://github.com/dotpcap/sharppcap/assets/9246717/9863d504-4a3f-4d50-a94c-c39557f3ac66)

**Reply by kayoub5** [reply]
> I have a "using" for SharpPcap.libpcap. I assumed that would provide libpcap from the SharpPcap assembly ?

no, sharppcap is just a wrapper around what's already installed on the system.
